ssl/statem/statem_clnt.c

```diff
@@ -1136,6 +1138,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
                session_id_len);
     }
 
+    /* Session version and negotiated protocol version should match
+     * (This may not be the case when reusing old session)
+     */
+    if (s->version != s->session->ssl_version) {
```

Comment:
The `if` above only sets it to 1 in the case of a session id match, i.e. a non-stale session?? The faulty server thing I get. That is, I'm not disputing the need for this check...only the comment that describes it (I'd suggest just deleting the second line of the comment).